src/parser.rs

```diff
@@ -4193,6 +4194,40 @@ impl<'a> Parser<'a> {
         }
     }
 
+    pub fn parse_set_operator_option(
```

Comment:
I'd be cautious about 2 things:
1. Are you really sure that the `ALL` and `DISTINCT` change between operations? 
2. You're being restrictive here. Please add `GenericDialect` and be aware of one thing: unless you're REALLY sure that only those 2 dialects support that feature, please be permissive and do not validate dialects.

I found that at least `ANSI` supports this. Please, or be permissive or validate perfectly (which is really hard)

Reply:
I think ALL / DISTINCT are ANSI SQL and thus should be handled by all dialects (unless we have specific evidence that some dialects don't support them there is no reason to prevent them being parsed)

Reply:
Thank you for commenting.

> Are you really sure that the ALL and DISTINCT change between operations?
sorry, could you elaborate this? 

> You're being restrictive here. Please add GenericDialect and be aware of one thing: unless you're REALLY sure that only those 2 dialects support that feature, please be permissive and do not validate dialects.

> I think ALL / DISTINCT are ANSI SQL and thus should be handled by all dialects (unless we have specific evidence that some dialects don't support them there is no reason to prevent them being parsed)

OK, I'll remove the specific dialects handling (bq, mysql) and make it parse by GenericDialect.

Reply:
@unvalley you're parsing them differently. `Union` is being parsed with both `ALL` and `DISTINCT`, while `Except` and `Intersect` not.

Is there any reason why you can't do something like:
```
match op {
            Some(SetOperator::Union) => {
                if self.parse_keyword(Keyword::ALL) {
                    Some(SetOperatorOption::All)
                } else if self.parse_keyword(Keyword::DISTINCT)
                    && dialect_of!(self is MySqlDialect | BigQueryDialect)
                {
                    Some(SetOperatorOption::Distinct)
                } else {
                    None
                }
            }
            Some(SetOperator::Except) | Some(SetOperator::Intersect) => {
                if self.parse_keyword(Keyword::ALL) {
                    Some(SetOperatorOption::All)
                } else {
                    None
                }
            }
            _ => None,
        }
```

Also, FYI (not saying you must do it in this PR, besides the second one):
- ANSI supports both [(1)](https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#query-expression-body). 
- The `GenericDialect` should be added to parse it as well. Unless the syntax is REALLY different between dialects, I'd say to add every syntax to the generic because, as the name states, it tries to be as generic as possible.
- MySQL also accepts it in `INTERSECT` [(2)](https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#query-expression-body) and `EXCEPT` [(3)](https://dev.mysql.com/doc/refman/8.0/en/except.html).
- BigQuery, following your implementation, support both only for `UNION` [(4)](https://dev.mysql.com/doc/refman/8.0/en/except.html)

[1] : https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#query-expression-body
[2] : https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#query-expression-body
[3] : https://dev.mysql.com/doc/refman/8.0/en/except.html
[4] : https://dev.mysql.com/doc/refman/8.0/en/except.html

Reply:
Thanks, I got it. I should implement `ALL` and `DISTINCT` for `INTERSECT` and `EXCEPT` as I do for `UNION`. And, I should focus mainly on `GenericDialect`.
In this pr, I'll only implement the [1] and [2] for `Generic Dialect`.

I thought the BigQuery dialect should not allow `ALL` to be attached to `INTERSECT` and `EXCEPT`, but I see that is not correct. Because ANSI supports it, there is no trouble with parsing being done. And it is difficult to test that it cannot be parsed as you said.

Reply:
@unvalley one option was to parse it conditionally for all of them, the same you do for `UNION`. But IMHO I think this can be scoped in another PR.

Your PR is to support `UNION DISTINCT` for both dialects and, parsing with `Generic`, you can have it.

If you want the EXACT same support for a dialect (being it permissive or restrictive), I'd say to do it separately, working from a generic scope to a restrict one. Trying to do all at once make the code confusing, and requires reading all documentations at once.